File: js/05_busca_sequencial.js
// Lista de números
const nums = [9, 21, 33, 12, 0, 18, 24, 30, 15, 6, 3, 27];

/**
* Função que realiza uma busca sequencial em um array procurando por um determinado valor.
* Se o valor for encontrado, retorna a posição onde foi encontrado.
* Caso contrário, retorna o valor -1.
*/
function buscaSequencial(array, valor) {
    for (let i = 0; i < array.length; i++) {
        // Encontrou o valor; retorna a posição onde foi encontrado
        if (valor === array[i]) {
            return i;
        }
    }
    // Percorreu todo o array e não encontrou o valor: retorna -1
    return -1;
}

// Busca sequencial: Lista precisa estar ordenada e quanto mais perto do fim o item estiver, mais tempo o algoritmo demora para encontrar
// Só é eficiente se o item procurado estiver no começo da lista


// TESTES

// Procurando o valor 15 no array nums
let resultado = buscaSequencial(nums, 15);
console.log(`Posição do valor 15 no array: ${resultado}`);
console.log("\n" + "-".repeat(80) + "\n");

// Procurando o valor 20 no array nums
resultado = buscaSequencial(nums, 20);
console.log(`Posição do valor 20 no array: ${resultado}`);
console.log("\n" + "-".repeat(80) + "\n");

// Procurando o valor 33 no array nums
resultado = buscaSequencial(nums, 33);
console.log(`Posição do valor 33 no array: ${resultado}`);
console.log("\n" + "-".repeat(80) + "\n");


// TESTES COM NOMES

let nomes = require('./data/nomes_desord.json');

let horaIni = Date.now();
resultado = buscaSequencial(nomes, 'JOAO');
let horaFim = Date.now();
console.log(`Posição do valor "JOAO" na lista: ${resultado}`);
console.log(`Tempo gasto: ${horaFim - horaIni}ms`);
console.log("\n" + "-".repeat(80) + "\n");


nomes = require('./data/lista_nomes.json');

horaIni = Date.now();
resultado = buscaSequencial(nomes, 'JOAO');
horaFim = Date.now();
console.log(`Posição do valor "JOAO" na lista: ${resultado}`);
console.log(`Tempo gasto: ${horaFim - horaIni}ms`);
console.log("\n" + "-".repeat(80) + "\n");

horaIni = Date.now();
resultado = buscaSequencial(nomes, 'CARLOS');
horaFim = Date.now();
console.log(`Posição do valor "CARLOS" na lista: ${resultado}`);
console.log(`Tempo gasto: ${horaFim - horaIni}ms`);
console.log("\n" + "-".repeat(80) + "\n");

horaIni = Date.now();
resultado = buscaSequencial(nomes, 'YARA');
horaFim = Date.now();
console.log(`Posição do valor "YARA" na lista: ${resultado}`);
console.log(`Tempo gasto: ${horaFim - horaIni}ms`);
console.log("\n" + "-".repeat(80) + "\n");

horaIni = Date.now();
resultado = buscaSequencial(nomes, 'ORKUTILSON');
horaFim = Date.now();
console.log(`Posição do valor "ORKUTILSON" na lista: ${resultado}`);
console.log(`Tempo gasto: ${horaFim - horaIni}ms`);
console.log("\n" + "-".repeat(80) + "\n");

File: python/05_busca_sequencial.py
# Lista de números
nums = [9, 21, 33, 12, 0, 18, 24, 30, 15, 6, 3, 27]

def busca_sequencial(lista, val):
    """ Função que realiza uma busca sequencial em uma lista procuração por val.
    Se val for encontrado, retorna a posição de val.
    Caso contrário, retorna o valor -1 """
    
    for pos in range(len(lista)):
        # Encontrou val; retorna a posição onde foi encontrado
        if val == lista[pos]: return pos
    # Percorreu toda a lista e não encontrou val: retorna -1
    return -1

# Busca sequencial: Lista precisa estar ordenada e quanto mais perto do fim o item estiver, mais tempo o algoritmo demora para encontrar
# Só é eficiente se o item procurado estiver no começo da lista


# TESTES

# Procurando o valor 15
resultado = busca_sequencial(nums, 15)
print(f'Posição do valor 15 na lista: {resultado}')
print('\n', '-' * 80, '\n', sep='')

# Procurando o valor 20
resultado = busca_sequencial(nums, 20)
print(f'Posição do valor 20 na lista: {resultado}')
print('\n', '-' * 80, '\n', sep='')

# Procurando o valor 33
resultado = busca_sequencial(nums, 33)
print(f'Posição do valor 33 na lista: {resultado}')
print('\n', '-' * 80, '\n', sep='')


# TESTES COM NOMES

import sys
sys.dont_write_bytecode = True # Impede a criação do cache

from time import time
from data.nomes_desord import nomes

hora_ini = time()
resultado = busca_sequencial(nomes, 'JOAO')
hora_fim = time()
print(f'Posição do valor "JOAO" na lista: {resultado}')
print(f'Tempo gasto: {round((hora_fim - hora_ini) * 1000, 2)}ms')
print('\n', '-' * 80, '\n', sep='')


from data.lista_nomes import nomes

hora_ini = time()
resultado = busca_sequencial(nomes, 'JOAO')
hora_fim = time()
print(f'Posição do valor "JOAO" na lista: {resultado}')
print(f'Tempo gasto: {round((hora_fim - hora_ini) * 1000, 2)}ms')
print('\n', '-' * 80, '\n', sep='')

hora_ini = time()
resultado = busca_sequencial(nomes, 'CARLOS')
hora_fim = time()
print(f'Posição do valor "CARLOS" na lista: {resultado}')
print(f'Tempo gasto: {round((hora_fim - hora_ini) * 1000, 2)}ms')
print('\n', '-' * 80, '\n', sep='')

hora_ini = time()
resultado = busca_sequencial(nomes, 'YARA')
hora_fim = time()
print(f'Posição do valor "YARA" na lista: {resultado}')
print(f'Tempo gasto: {round((hora_fim - hora_ini) * 1000, 2)}ms')
print('\n', '-' * 80, '\n', sep='')

hora_ini = time()
resultado = busca_sequencial(nomes, 'ORKUTILSON')
hora_fim = time()
print(f'Posição do valor "ORKUTILSON" na lista: {resultado}')
print(f'Tempo gasto: {round((hora_fim - hora_ini) * 1000, 2)}ms')
print('\n', '-' * 80, '\n', sep='')
